src/osd/scrubber/scrub_queue.h

```diff
@@ -0,0 +1,199 @@
+  using SchedulingQueue = std::deque<SchedEntry>;
+
+  std::ostream& gen_prefix(std::ostream& out) const;
+
+  // ///////////////////////////////////////////////////
+  // the ScrubQueueOps interface:
+
+  utime_t scrub_clock_now() const override;
+
+  Scrub::sched_conf_t populate_config_params(
+      const pool_opts_t& pool_conf) const override;
+
+  void remove_entry(spg_t pgid, scrub_level_t s_or_d) final;
+
+  void cp_and_queue_target(SchedEntry t) final;
+
+  bool queue_entries(spg_t pgid, SchedEntry shallow, SchedEntry deep) final;
+
+
+  // ///////////////////////////////////////////////////
+  // outside the scope of the I/F used by the ScrubJob:
+
+  /**
+   * the main entry point for the OSD. Called in OSD::tick_without_osd_lock()
+   * to determine if there are PGs that are ready to be scrubbed, and to
+   * initiate a scrub of one of those that are ready.
+   */
+  void sched_scrub(
+      const ceph::common::ConfigProxy& config,
+      bool is_recovery_active);
+
+  /*
+   * handles a change to the configuration parameters affecting the scheduling
+   * of scrubs.
+   */
+  void on_config_times_change();
+
+ public:
+  void dump_scrubs(ceph::Formatter* f);
+
+  /**
+   * No new scrub session will start while a scrub was initiated on a PG,
+   * and that PG is trying to acquire replica resources.
+   */
+  void set_reserving_now() { a_pg_is_reserving = true; }
+  void clear_reserving_now() { a_pg_is_reserving = false; }
+  bool is_reserving_now() const { return a_pg_is_reserving; }
+
+  // resource reservation management
+
+ private:
+  bool can_inc_scrubs() const;
+
+ public:
+  bool inc_scrubs_local();
+  void dec_scrubs_local();
+  bool inc_scrubs_remote();
+  void dec_scrubs_remote();
+  void dump_scrub_reservations(ceph::Formatter* f) const;
+
+
+  /// counting the number of PGs stuck while scrubbing, waiting for objects
+  void mark_pg_scrub_blocked(spg_t blocked_pg);
+  void clear_pg_scrub_blocked(spg_t blocked_pg);
+
+ private:
+  int get_blocked_pgs_count() const;
+
+ public:
+  /**
+   * Pacing the scrub operation by inserting delays (mostly between chunks)
+   *
+   * Special handling for regular scrubs that continued into "no scrub" times.
+   * Scrubbing will continue, but the delays will be controlled by a separate
+   * (read - with higher value) configuration element
+   * (osd_scrub_extended_sleep).
+   */
+  std::chrono::milliseconds required_sleep_time(bool high_priority_scrub) const;
+
+  /**
+   *  called every heartbeat to update the "daily" load average
+   *
+   *  @returns a load value for the logger
+   */
+  [[nodiscard]] std::optional<double> update_load_average();
+
+ private:
+  CephContext* cct;
+  Scrub::ScrubSchedListener& osd_service;
+
+#ifdef WITH_SEASTAR
+  auto& conf() const
+  {
+    return local_conf();
+  }
+#else
+  auto& conf() const
+  {
+    return cct->_conf;
+  }
+#endif
+
+  mutable ceph::mutex jobs_lock = ceph::make_mutex("ScrubQueue::jobs_lock");
```

Comment:
This will need to be removed to reuse ScrubQueue within crimson.  Please move the lock to the OSDService and add an appropriate with_scrub_queue accessor.